### PR TITLE
ERC721Permit

### LINKED ISSUE #167: [posm]: Add Permit functionality to ERC721
### Component

Position Manager

### Describe the suggested feature and problem it solves.

Permit should: 
- [x] Use unordered nonces
- [x] Not be payable (in V3 it is payable)
- [ ] Use relevant libraries and base contracts (ie a PermitHash [library](https://github.com/Uniswap/v4-periphery/pull/153#discussion_r1681402782)) and a EIP1271 base contract that caches the domain separator. (see comments [here](https://github.com/Uniswap/v4-periphery/pull/153#discussion_r1681374875).
- [x] Constants above functions
- [x] Test: Ensure PERMIT_TYPEHASH is equal to expected hash
- [x] Use custom errors
- [x] replace magic constant with IERC1271.isValidSignature.selector
- [x] [this](https://github.com/Uniswap/v4-periphery/pull/141#discussion_r1684211300) comment

Ensure that all comments related to permit from [this PR](https://github.com/Uniswap/v4-periphery/pull/153) and [this PR](https://github.com/Uniswap/v4-periphery/pull/141) are resolved in the final feature PR.

### Describe the desired implementation.

_No response_

### Describe alternatives.

_No response_

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by saucepoint**
## Related Issue
Closes #167 and replaces https://github.com/Uniswap/v4-periphery/pull/205

## Description of changes

Adds ERC721Permit with

- [x] Solmate ERC721
- [x] OZ EIP712.sol
- [x] Unordered Nonces

### CODE REVIEW

File: src/base/ERC721Permit.sol
Comment (saucepoint) [inline src/base/ERC721Permit.sol:87]:
```suggestion
        digest =  _hashTypedDataV4(keccak256(abi.encode(PERMIT_TYPEHASH, spender, tokenId, _nonce, deadline)));
```

using OZ's EIP712 internal function..

File: src/base/ERC721Permit.sol
Comment (hensha256) [inline src/base/ERC721Permit.sol:62]:
```suggestion
        // override Solmate's ERC721 approve so approve() and permit() share the _approve method
```

File: src/PositionManager.sol
Comment (hensha256) [inline src/PositionManager.sol:62]: given this wont be implemented before audits i think its worth instead putting it in the constructor parameters? or a clearer comment about when it will be implemented?

File: src/base/ERC721Permit.sol
Comment (hensha256) [inline src/base/ERC721Permit.sol:61]: natspec?

File: src/base/ERC721Permit.sol
Comment (hensha256) [reply src/base/ERC721Permit.sol:62]: I'd also put this ocmment above the function as an `@ dev`

File: src/base/UnorderedNonce.sol
Comment (hensha256) [inline src/base/UnorderedNonce.sol:4]: natspec

File: test/shared/LiquidityOperations.sol
Comment (hensha256) [inline test/shared/LiquidityOperations.sol:96]: i think just rename! I had a similar thing with routing so i just had 1 file called `RoutingTestHelpers` i think do something similar! `PosmTestHelpers`?

File: test/shared/LiquidityOperations.sol
Comment (hensha256) [inline test/shared/LiquidityOperations.sol:102]: imo the whole point in a permit is that the signer often isnt the person who sends it. I think you should have a parameter of this function `caller` and do `vm.prank(caller)`. Then some tests you can pass in `signer` and others you can explicitly test that someone else can pass it in

Comment (hensha256) [reply test/shared/LiquidityOperations.sol:102]: e.g. a test where the `caller` is the `operator`!

File: test/position-managers/PositionManager.multicall.t.sol
Comment (hensha256) [inline test/position-managers/PositionManager.multicall.t.sol:104]:
```suggestion
        bytes memory actions = getDecreaseEncoded(tokenId, config, liquidityToRemove, ZERO_BYTES);
```

File: test/shared/LiquidityOperations.sol
Comment (hensha256) [reply test/shared/LiquidityOperations.sol:102]: ok i see youre already testing that - just not in this helper function. Thats fair. I _think_ I'd still rather this was `operator` not `signer` pranking just because the `signer` wouldnt do this - they'd just call `approve` instead

File: test/position-managers/Permit.t.sol
Comment (hensha256) [inline test/position-managers/Permit.t.sol:60]: would love a similar test to this but for the domain separator

File: test/position-managers/Permit.t.sol
Comment (hensha256) [inline test/position-managers/Permit.t.sol:103]: I feel like this file is missing specific tests of _permit_. Its testing "if you permit, X action works", but i feel like the actual state changes of permit are missing? Like a test that does a permit and checks

- The nonce wasnt set before, and then is set afterwards
- Bob wasnt approved, and then is afterwards
- Bob didnt get approved for all
- if you do another permit afterwards the old approval is wiped

You've overwritten the `approve` function, so that function needs testing too. Or if youd rather keep it simple you could remove the override of `approve` and just use the `_approve` function for `permit`

File: src/base/UnorderedNonce.sol
Comment (hensha256) [inline src/base/UnorderedNonce.sol:12]: would love tests of this function in particular - does it only set the bit in question, does it revert if the nonce was already taken?
if you look in core at `test_fuzz_flipTick`, something like that to check that `_useUnorderedNonce` always sets the right bit, doesnt set others, and reverts always when it was already set?

**Review by hensha256 — CHANGES_REQUESTED** [review]
Looks great! Mostly just want more testing!

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [reply src/base/UnorderedNonce.sol:12]: I think there are these exact fuzz tests in permit2 if you want to also transfer them over @saucepoint

File: src/base/ERC721Permit.sol
Comment (snreynolds) [inline src/base/ERC721Permit.sol:43]: I think we could really improve readability here a lot by moving this logic into a signature verification library..

really all this should look like in this code path is

```
bytes memory signature = ...
bytes32 permitHash = Permit.hash(spender, tokenId, nonce, deadline)
signature.verify(_hashTypedDataV4(permitHash), owner);  // also does all the support for IERC1271, and the ecrecover logic etc..
```

can reference also the permit2 signature library as well should be pretty similar

File: test/position-managers/Permit.t.sol
Comment (snreynolds) [inline test/position-managers/Permit.t.sol:24]: can we call this PermitERC721 ? Cause we will also have a permit for permit2 exposed

File: test/position-managers/Permit.t.sol
Comment (snreynolds) [inline test/position-managers/Permit.t.sol:25]: & also would be great to have a suite of unit tests for approve

File: test/position-managers/Permit.t.sol
Comment (saucepoint) [reply test/position-managers/Permit.t.sol:24]: will rename this at the end before merge to preserve reviews / diffs

File: src/base/ERC721Permit.sol
Comment (snreynolds) [inline src/base/ERC721Permit.sol:20]: do we need to declare this here? don't see it used internally

File: src/base/ERC721Permit.sol
Comment (snreynolds) [inline src/base/ERC721Permit.sol:6]: remove

File: src/base/ERC721Permit.sol
Comment (snreynolds) [inline src/base/ERC721Permit.sol:19]: random space?

File: src/base/ERC721Permit.sol
Comment (snreynolds) [inline src/base/ERC721Permit.sol:29]: imo this should live in the `EIP712` contract

File: src/base/ERC721Permit.sol
Comment (snreynolds) [inline src/base/ERC721Permit.sol:56]: inheritdoc?

**Review by snreynolds — COMMENTED** [review]
ok thoughts on not using openzeppelin contracts at all here.. i think so far its now just EIP712 which gives me a headache looking at... can we just replace w our own?

File: src/base/ERC721Permit.sol
Comment (saucepoint) [reply src/base/ERC721Permit.sol:20]: it was part of IERC721Permit, from v3

not sure if backend/interface depends on it 🤔

Comment (hensha256) [reply src/base/ERC721Permit.sol:20]: i feel like its not worth an rpc call to look this up we can just calculate it offchain? i say remove

File: src/base/UnorderedNonce.sol
Comment (hensha256) [inline src/base/UnorderedNonce.sol:16]: technically we have a gas optimised version of this in core if it interests you
https://github.com/Uniswap/v4-core/blob/799dd2cb980319a8d3b827b6a7aa59a606634553/src/libraries/TickBitmap.sol#L48-L52

File: test/ERC721Permit.t.sol
Comment (hensha256) [inline test/ERC721Permit.t.sol:54]: these tests al need `_fuzz` in the name

File: test/UnorderedNonce.t.sol
Comment (hensha256) [inline test/UnorderedNonce.t.sol:99]: i feel like this test doesnt actually check the unordered nonce contract? it just tests the `invalidateUnorderedNonces` function which doesnt use UnorderedNonce it just implements something custom?

e.g. there can easily be overlap between `startBitmap` and `mask` having the same bit set... and that doesnt revert because its not actually using the nonce checking.

Honestly i just think this test and the one above can be deleted? The test `testInvalidateNonzeroWord` had a nice use of this function to then check that used nonces reverted

**Review by hensha256 — COMMENTED** [review]
lots of really nice testing in here 👯

File: src/base/UnorderedNonce.sol
Comment (saucepoint) [reply src/base/UnorderedNonce.sol]: took at crack at this, so please review closely ❤️

File: src/base/ERC721Permit.sol
Comment (saucepoint) [reply src/base/ERC721Permit.sol]: changed it to external, only reason why this (and PERMIT_TYPEHASH) were public because of IERC721Permit

if we're in agreement, I can change the underlying interface

File: test/mocks/MockUnorderedNonce.sol
Comment (hensha256) [reply test/mocks/MockUnorderedNonce.sol:12]: wait sorry i meant _this_ function feels like a `batchSpendNonces`? the one above is a `spendNonce` it just uses one right?

File: src/interfaces/IERC721Permit.sol
Comment (hensha256) [inline src/interfaces/IERC721Permit.sol:20]: maybe worth an @dev to say why its payable lol

Comment (hensha256) [reply src/interfaces/IERC721Permit.sol:20]: sorry again kenny

File: src/base/UnorderedNonce.sol
Comment (hensha256) [reply src/base/UnorderedNonce.sol]: it looks great!!! 👯

File: src/interfaces/IERC721Permit.sol
Comment (snreynolds) [reply src/interfaces/IERC721Permit.sol:20]: wait I added this in my other PR

Comment (snreynolds) [reply src/interfaces/IERC721Permit.sol:20]: where u left a similar comment haah, so i got this one

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [inline src/base/UnorderedNonce.sol:46]: if we're going to use .revertWith lib from core this will need to be changed

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [inline src/base/UnorderedNonce.sol:20]: imo this does more than flip the bit so we should maybe split out this code.

1 - use the nonce to extract the wordPos/the bitmap to operate on 
2- flip bit

flipBit(nonce) makes it seem like the nonce thats being passed in is the bit thats being flipped which is not really how the data is used

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [inline src/base/UnorderedNonce.sol:21]: can we fuzz test equivalence? this is a lot of assembly :P

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [reply src/base/UnorderedNonce.sol:46]: are we doing that? lol

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:46]: we can put that in the error pr just a call out tho

File: test/mocks/MockUnorderedNonce.sol
Comment (saucepoint) [reply test/mocks/MockUnorderedNonce.sol:12]: oh lol makes sense. got confused since spendNonce can technically "batch" spend multiple bits

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [inline src/base/UnorderedNonce.sol:18]: is it more expensive if we did nonces[owner][wordPos].flipBit(bitPos)? can you see? i think i just like that abstraction a little better

Comment (saucepoint) [reply src/base/UnorderedNonce.sol:18]: its not as friendly since `flipBit` would calculate the new bitmap and return it so the caller can set it in storage

the alternative/cleaner syntax might be `nonces[owner].spendNonce(nonce)` where `spendNonce` does the wordPos/bitPos logic

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:18]: hmmm

Comment (hensha256) [reply src/base/UnorderedNonce.sol:18]: > nonces[owner][wordPos].flipBit(bitPos)

yeah generally nested mappings are much more expensive becaus eyou do more hashing (and hashing is on the verge of getting even more expensive, theres a change coming)

however i like `nonces[owner].spendNonce(nonce)` this!!

Comment (hensha256) [reply src/base/UnorderedNonce.sol:18]: ok i misunderstood - now i see it lol
i agree that doing it on the word would be messy becaus eyou have to write
```
nonces[owner][wordPos] = nonces[owner][wordPos].flipBit(x)
```

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:18]: i think i still prefer the abstraction where we were to change it to 

nonces[owner][wordPos].flipTick bc the functions become more reusable (to extract bitPos and wordPos) from nonce, then to get the new bitmap...

buttt i get it if its too expensive

maybe we could still have nonces[owner].spendNonce() that still use other helpers for those two things? im just noticing that we have to rewrite helpers in tests to do the same thing weve written here/ I could see those functions being helpful generally in a library

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:18]: Can you see how much gas is saved not doing it this way? @saucepoint 

like is it significantly more expsenive to do

nonces[owner][wordPos] = nonces[owner][wordPos].flipTick(bitPos) ?

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:18]: i will also say tho.. this is not a super heavily used codepath so we could just go back to the solidity version and be done w it, theres no need to hyper optimize this path

File: src/base/ERC721Permit.sol
Comment (saucepoint) [reply src/base/ERC721Permit.sol]: note for self: leaving DOMAIN_SEPARATOR and will move to EIP712.sol if/when its written

File: src/base/UnorderedNonce.sol
Comment (hensha256) [reply src/base/UnorderedNonce.sol:18]: yeah im happy with that 🤷 no strong opinions

File: src/base/UnorderedNonce.sol
Comment (snreynolds) [inline src/base/UnorderedNonce.sol:9]: natspec that the word is actually max type(uint248).max

File: test/position-managers/Permit.t.sol
Comment (snreynolds) [inline test/position-managers/Permit.t.sol:76]: you dont actually need permissions to increase, so can probably remove this?

Comment (snreynolds) [reply test/position-managers/Permit.t.sol:76]: or you can leave it to show you dont need a permit

File: test/position-managers/Permit.t.sol
Comment (snreynolds) [inline test/position-managers/Permit.t.sol:109]: can you add an expect revert bob trying to adjust alices position before the permit

Comment (saucepoint) [reply test/position-managers/Permit.t.sol:109]: tested in `test_noPermit_decreaseLiquidityRevert`

File: test/position-managers/Permit.t.sol
Comment (saucepoint) [reply test/position-managers/Permit.t.sol:76]: removing

the test `test_increaseLiquidity_withUnapprovedCaller` shows that users can increase without approval

**Review by snreynolds — APPROVED** [review]

**Pull request merged**
